src/stylus/StylusDeployer.sol

```diff
@@ -0,0 +1,177 @@
+    /// @param initData Initialisation call data. After deployment the contract will be called with this data
+    ///                 If no initialisation data is provided then the newly deployed contract will not be called.
+    ///                 This means that the fallback function cannot be called from this contract, only named functions can be.
+    /// @param initValue Initialisation value. After deployment, the contract will be called with the init data and this value.
+    ///                  At least as much eth as init value must be provided with this call. Init value is specified here separately
+    ///                  rather than using the msg.value since the msg.value may need to be greater than the init value to accomodate activation data fee.
+    ///                  See the @notice block above for more details.
+    /// @param salt If a non zero salt is provided the contract will be created using CREATE2 instead of CREATE
+    ///             The supplied salt will be hashed with the initData and the initValue so that wherever the address is observed
+    ///             it was initialised with the same variables.
+    /// @return The address of the deployed conract
+    function deploy(
+        bytes calldata bytecode,
+        bytes calldata initData,
+        uint256 initValue,
+        bytes32 salt
+    ) public payable returns (address) {
+        if (salt != 0) {
+            // if a salt was supplied, hash the salt with init value and init data. This guarantees that
+            // anywhere the address of this contract is seen the same init data and value were used.
+            salt = initSalt(salt, initData, initValue);
+        }
+
+        address newContractAddress = deployContract(bytecode, salt);
+        bool shouldActivate = requiresActivation(newContractAddress);
+        if (shouldActivate) {
+            // ensure there will be enough left over for init
+            uint256 activationValue = msg.value - initValue;
+            ARB_WASM.activateProgram{value: activationValue}(newContractAddress);
+        }
+
+        // initialize - this will fail if the program wasn't activated by this point
+        // we check if initData exists to avoid calling contracts unnecessarily
+        if (initData.length != 0) {
+            (bool success,) = address(newContractAddress).call{value: initValue}(initData);
+            if (!success) {
+                revert ContractInitializationError(newContractAddress);
+            }
+        } else if (initValue != 0) {
+            // if initValue exists init data should too
+            revert InitValueButNotInitData();
+        }
+
+        // refund any remaining value if:
+        // - activation can return some to this contract
+        // - some activation value was supplied but not used
+        if (shouldActivate || msg.value != initValue) {
+            uint256 bal = address(this).balance;
```

Comment:
This is not safe against reentrancy, the inner deployment will get all the refund. We can instead check pre and post balance.

Reply:
Now using:
```
uint256 bal = msg.value - dataFee - initValue;
```